stdlib/LinearAlgebra/src/special.jl

```diff
@@ -148,3 +148,119 @@ function fill!(A::Union{Diagonal,Bidiagonal,Tridiagonal,SymTridiagonal}, x)
     throw(ArgumentError("array of type $(typeof(A)) and size $(size(A)) can
     not be filled with $x, since some of its entries are constrained."))
 end
+
+# equals and approx equals methods for structured matrices
+# SymTridiagonal == Tridiagonal is already defined in tridiag.jl
+
+# SymTridiagonal and Bidiagonal have the same field names
+==(A::Diagonal, B::Union{SymTridiagonal, Bidiagonal}) = iszero(B.ev) && A.diag == B.dv
+==(B::Bidiagonal, A::Diagonal) = A == B
+
+==(A::Diagonal, B::Tridiagonal) = iszero(B.dl) && iszero(B.du) && A.diag == B.d
+==(B::Tridiagonal, A::Diagonal) = A == B
+
+function ==(A::Bidiagonal, B::Tridiagonal)
+    if A.uplo == 'U'
+        return iszero(B.dl) && A.dv == B.d && A.ev == B.du
+    else
+        return iszero(B.du) && A.dv == B.d && A.ev == B.dl
+    end
+end
+==(B::Tridiagonal, A::Bidiagonal) = A == B
+
+==(A::Bidiagonal, B::SymTridiagonal) = iszero(B.ev) && iszero(A.ev) && A.dv == B.dv
+==(B::SymTridiagonal, A::Bidiagonal) = A == B
+
+function isapprox(A::Diagonal, B::Diagonal;
+    atol::Real=0,
+    rtol::Real=Base.rtoldefault(promote_leaf_eltypes(A),promote_leaf_eltypes(B),atol),
+    nans::Bool=false, norm::Function=norm)
+
+    return isapprox(A.diag, B.diag; rtol=rtol, atol=atol, nans=nans)
+end
+
+function isapprox(A::Diagonal, B::Union{Bidiagonal, SymTridiagonal};
+    atol::Real=0,
+    rtol::Real=Base.rtoldefault(promote_leaf_eltypes(A),promote_leaf_eltypes(B),atol),
+    nans::Bool=false, norm::Function=norm)
+
+    return iszero(B.ev) && isapprox(A.diag, B.dv; rtol=rtol, atol=atol, nans=nans)
```

Comment:
Should this be a test for approximately zero when atol!=0?

Reply:
The norm argument is unused

Reply:
For `vecnorm` it would be `hypot(vecnorm(B.ev), vecnorm(A.diag .- B.dv)) <= ...`, but to deal with the general case I guess you need a fallback.

Reply:
`vecnorm` was deprecated with v0.7. The problem with using a recursive `isapprox` fallback is that `isapprox` doesn't have a kwarg for `norm` when the eltype is `Number`.